devtools/dbt_helper.py

```diff
@@ -0,0 +1,296 @@
+        return [
+            {
+                "check_row_counts_per_partition": {
+                    "table_name": table_name,
+                    "partition_column": partition_column,
+                }
+            }
+        ]
+
+    @classmethod
+    def from_table_name(cls, table_name: str, partition_column: str) -> "DbtSchema":
+        """Construct configuration defining table from PUDL metadata."""
+        return cls(
+            data_tests=cls.get_row_count_test_dict(table_name, partition_column),
+            name=table_name,
+            columns=[
+                DbtColumn(name=f.name)
+                for f in PUDL_PACKAGE.get_resource(table_name).schema.fields
+            ],
+        )
+
+
+class DbtSource(BaseModel):
+    """Define basic dbt yml structure to add a pudl table as a dbt source."""
+
+    name: str = "pudl"
+    tables: list[DbtTable]
+
+
+class DbtSchema(BaseModel):
+    """Define basic structure of a dbt models yaml file."""
+
+    version: int = 2
+    sources: list[DbtSource]
+
+    @classmethod
+    def from_table_name(cls, table_name: str, partition_column: str) -> "DbtSchema":
+        """Construct configuration defining table from PUDL metadata."""
+        return cls(
+            sources=[
+                DbtSource(
+                    tables=[DbtTable.from_table_name(table_name, partition_column)]
+                )
+            ]
+        )
+
+
+def get_data_source(table_name: str) -> str:
+    """Return data source for a table or 'output' if there's more than one source."""
+    resource = PUDL_PACKAGE.get_resource(table_name)
+    if len(resource.sources) > 1:
+        return "output"
+
+    return resource.sources[0].name
+
+
+AddTableResult = namedtuple("AddTableResult", ["success", "message"])
+
+
+def _get_nightly_url(table_name: str) -> str:
+    return f"https://s3.us-west-2.amazonaws.com/pudl.catalyst.coop/nightly/{table_name}.parquet"
+
+
+def _get_model_path(table_name: str, data_source: str) -> Path:
+    return Path("./dbt") / "models" / data_source / table_name
+
+
+def _get_row_count_csv_path() -> Path:
+    return Path("./dbt") / "seeds" / "row_counts.csv"
+
+
+def generate_row_counts(
+    table_name: str,
+    partition_column: str = "report_year",
+    use_local_tables: bool = False,
+    clobber: bool = False,
+) -> AddTableResult:
+    """Generate row counts per partition and write to csv file within dbt project."""
+    # Get existing row counts table
+    row_counts_df = pd.read_csv(_get_row_count_csv_path())
+
+    if table_name in row_counts_df["table_name"].to_numpy() and not clobber:
+        return AddTableResult(
+            success=False,
+            message=f"There are already row counts for table {table_name} in row counts table and clobber is not set.",
+        )
+
+    # Load table of interest
+    if not use_local_tables:
+        df = pd.read_parquet(_get_nightly_url(table_name))
```

Comment:
Rather than downloading the entire parquet file to get the row counts per partition (which will mean several GB every time this is run on all tables) can we potentially send a query that only has to access the Parquet metadata? That would require almost no data transfer, but might mean we need to change how the parquet files are indexed and possibly use `s3://` rather than `https://` URLs to access the nightly outputs.

Reply:
I changed this to use `duckdb` to query the nightly parquet files and I've observed significantly reduced network usage when running the script, so I don't think it's downloading the full table anymore.